packages/block-library/src/heading/index.php

```diff
@@ -0,0 +1,98 @@
+<?php
+/**
+ * Server-side rendering of the `core/heading` block.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Adds a wp-block-heading class to the heading blocks without storing
+ * it in the serialized $content.
+ *
+ * @param array  $attributes Attributes of the block being rendered.
+ * @param string $content Content of the block being rendered.
+ *
+ * @return string The content of the block being rendered.
+ */
+function block_core_heading_render( $attributes, $content ) {
+	if ( ! $content ) {
+		return $content;
+	}
+	$matches = array();
+	$pattern = '/
+        ^(\s*)                               # Any leading whitespace.
+        <(?<tag_name>
+            h[1-6]                           # The opening tag...
+            (?=\s|>)                         # ...followed by a whitespace or >
+                                             # ?= means a "lookahead"
+        )
+        (?<before_class>                     # Any attributes prior to "class"
+            (?:                              # ?: is a "non-capturing group"
+                (?!\sclass=["\'])[^>]        # Match all characters until ">" except when
```

Comment:
these negative lookahead can behave very slowly in practice, particularly if we end up in backtracking situations. ungreedy matches can help us because they let us express everything _until_ this part that we're looking for but don't incur the same performance issues that negative lookahead do

Reply:
I know how to do negative non-greedy matching with character classes, but how would you do it with a word?

Reply:
Nevermind, it clicked the moment I pressed "submit".